pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -84,21 +85,20 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
 	// get the current, and update revisions
 	currentRevision, updateRevision, collisionCount, err := ssc.getStatefulSetRevisions(set, revisions)
 	if err != nil {
-		return err
+		return multierr.Append(err, ssc.truncateHistory(set, pods, revisions, currentRevision, updateRevision))
```

Comment:
I think this works, but would it be better if we split this into two functions so we only need to add the `truncatehistory` call in one place?

Reply:
Do you mean creating a new function which invokes `truncateHistory` first and then merges the errors?

Reply:
Like this:

```
func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, pods []*v1.Pod) error {

	// list all revisions and sort them
	revisions, err := ssc.ListRevisions(set)
	if err != nil {
		return err
	}
	history.SortControllerRevisions(revisions)

	// get the current, and update revisions
	currentRevision, updateRevision, collisionCount, err := ssc.getStatefulSetRevisions(set, revisions)
	if err != nil {
		return handleCleanup(err, ssc, set, pods, revisions, currentRevision, updateRevision)
	}

	// perform the main update function and get the status
	status, err := ssc.updateStatefulSet(set, currentRevision, updateRevision, collisionCount, pods)
	if err != nil {
		return handleCleanup(err, ssc, set, pods, revisions, currentRevision, updateRevision)
	}

	// update the set's status
	err = ssc.updateStatefulSetStatus(set, status)
	if err != nil {
		return handleCleanup(err, ssc, set, pods, revisions, currentRevision, updateRevision)
	}
	klog.V(4).Infof("StatefulSet %s/%s pod status replicas=%d ready=%d current=%d updated=%d",
		set.Namespace,
		set.Name,
		status.Replicas,
		status.ReadyReplicas,
		status.CurrentReplicas,
		status.UpdatedReplicas)

	klog.V(4).Infof("StatefulSet %s/%s revisions current=%s update=%s",
		set.Namespace,
		set.Name,
		status.CurrentRevision,
		status.UpdateRevision)

	// maintain the set's revision history limit
	return ssc.truncateHistory(set, pods, revisions, currentRevision, updateRevision)
}

func handleCleanup(
	err error,
	ssc *defaultStatefulSetControl,
	set *apps.StatefulSet,
	pods []*v1.Pod,
	revisions []*apps.ControllerRevision,
	currentRevision *apps.ControllerRevision,
	updateRevision *apps.ControllerRevision) error {
	return multierr.Append(err, ssc.truncateHistory(set, pods, revisions, currentRevision, updateRevision))
}
```